### PR TITLE
db: system_keyspace: take the reserved_memory into account

### LINKED ISSUE #12522: system_tables_test: test_default_content: AssertionError: Unexpected memory value: 1019215872
Seen first in https://jenkins.scylladb.com/view/master/job/scylla-master/job/dtest-daily-release/175/testReport/system_tables_test/TestRuntimeInfoTable/Run_Dtest_Parallel_Cloud_Machines___FullDtest___full_split000___test_default_content/
Scylla version abc43f97c916
```
AssertionError: Unexpected memory value: 1019215872
assert 1019215872 == 1073741824
  +1019215872
  -1073741824
```

### PR DESCRIPTION
**Pull request by tchaikov**
before this change, we returns the total memory managed by Seastar in the "total" field in system.memory. but this value only reflect the total memory managed by Seastar's allocator. if `reserve_additional_memory` is set when starting app_template, Seastar's memory subsystem just reserves a chunk of memory of this specified size for system, and takes the remaining memory. since f05d612da8, we set this value to 50MB for wasmtime runtime. hence the test of `TestRuntimeInfoTable.test_default_content` in dtest fails. the test expects the size passed via the option of `--memory` to be identical to the value reported by system.memory's "total" field.

after this change, the "total" field takes the reserved memory for wasm udf into account. the "total" field should reflect the total size of memory used by Scylla, no matter how we use a certain portion of the allocated memory.

Fixes #12522
Signed-off-by: Kefu Chai <kefu.chai@scylladb.com>

### CODE REVIEW

**Comment by tchaikov** [comment]
see also previous discussion at https://github.com/scylladb/scylla-dtest/pull/2991

**Comment by tchaikov** [comment]
@bhalevy hi Benny, after a second thought. i changed my mind. and i think "total" should reflect the total size of used memory no matter what is used for, either seastar or wasm udf. so in this change, i just sum them up and return using the old "total" field. i will create a dedicated PR to expose the memory reserved for "wasm_udf" if it still relevant. but i doubt it, as it looks kind of weird sitting aside of "total", "used" and "free". as it is a domain specific value.

File: db/system_keyspace.cc
Comment (bhalevy) [inline db/system_keyspace.cc:2282]: Is it reserved per shard?

**Comment by bhalevy** [comment]
> @bhalevy hi Benny, after a second thought. i changed my mind. and i think "total" should reflect the total size of used memory no matter what is used for, either seastar or wasm udf.

make sense

> so in this change, i just sum them up and return using the old "total" field. i will create a dedicated PR to expose the memory reserved for "wasm_udf" if it still relevant. but i doubt it, as it looks kind of weird sitting aside of "total", "used" and "free". as it is a domain specific value.

I agree.  I think exposing it via a metric would be more appropriate.

File: db/system_keyspace.cc
Comment (tchaikov) [reply db/system_keyspace.cc:2282]: indeed. will fix.

**Comment by tchaikov** [comment]
changelog

- the init value for calculating the total memory should be multiplied by `smp::count`, as the reserved memory option is applied to every shard.

**Comment by tchaikov** [comment]
> > so in this change, i just sum them up and return using the old "total" field. i will create a dedicated PR to expose the memory reserved for "wasm_udf" if it still relevant. but i doubt it, as it looks kind of weird sitting aside of "total", "used" and "free". as it is a domain specific value.
> 
> I agree. I think exposing it via a metric would be more appropriate.
> 
> If exposed in the system table, I'd expose it in a generic way like "reserved", without mentioning it's for wasm, so in the future it could contain other reservations (that would be out of the "used" count)

sure. i will create a follow-up once this once gets reviewed, approved and merged. but i wonder how user should interpret this value without more context?

File: db/system_keyspace.cc
Comment (bhalevy) [reply db/system_keyspace.cc]: I'm confused. If total is passed to map_reduce so it would be summed smp::count times, so I think the initialization was correct.

My concern was that the 50MB value was for all shards (like the totla memory option is)

Comment (tchaikov) [reply db/system_keyspace.cc]: sorry, i misinterpreted https://github.com/scylladb/seastar/blob/b5865a96039243c74db8d1e21f21099dde3028fe/src/core/resource.cc#L512 . 50MB is reserved from **total** memory, i.e., the "1G" in `--memory 1G`.

Comment (tchaikov) [reply db/system_keyspace.cc:2282]: >  If total is passed to map_reduce so it would be summed smp::count times, so I think the initialization was correct.

no, how so? it will be something like
```py
total = db::config::wasm_udf_reserved_memory + stats_on_shard[0].total + stats_on_shard[1].total + ...
```

so the "total" is correct. i am not sure if i am following you now.

Comment (bhalevy) [reply db/system_keyspace.cc:2282]: math is hard ;)
I trust your calculation.

**Review by bhalevy — APPROVED** [review]

**Review by wmitros — APPROVED** [review]
LGTM, although now that I look at this, I might have set the value of 50MB incorrectly - it should have probably been 50MB per shard, but as you've noticed, it's 50MB in total - I'll make a separate issue. It shouldn't conflict with this patch, because even if we change `db::config::wasm_udf_reserved_memory` later, the algorithm will still work

**Comment by tchaikov** [comment]
@bhalevy hi Benny, am i supposed to ping a maintainer to include this change in the next branch, so we are able to have it tested and promoted if it survives the tests?

**Comment by denesb** [comment]
> @bhalevy hi Benny, am i supposed to ping a maintainer to include this change in the next branch, so we are able to have it tested and promoted if it survives the tests?

Queued.

Feel free to ping @scylladb/scylla-maint when a PR is ready-to-merge (all reviewers approved and CI is passed). Some people wait for some time for maintainers discover mergeable PRs on their own, others ping immediately. Both are valid approaches. If there is a maintainer among the reviewers it is them who merges the PR eventually, but this is in no way a rule.

**Comment by tchaikov** [comment]
> LGTM, although now that I look at this, I might have set the value of 50MB incorrectly - it should have probably been 50MB per shard, but as you've noticed, it's 50MB in total - I'll make a separate issue. It shouldn't conflict with this patch, because even if we change `db::config::wasm_udf_reserved_memory` later, the algorithm will still work

@wmitros hi Wojciech, now that this change has been merged. shall we add the change you proposed? or i can take care of it? it might be a good opportunity to try out https://github.com/scylladb/seastar/blob/e0011bf957baaf44efc45930c87d1f4395d3d733/include/seastar/core/units.hh#L33 =)

**Comment by wmitros** [comment]
@tchaikov hi Kefu, I agree that we should proceed with fixing this.

I was considering fixing this by sharing the amount of reserved memory specified by `wasm_udf_reserved_memory` between shards. I didn't want to simply divide the memory between shards, because we don't know the number of shards when setting `app_cfg.reserve_additional_memory` yet, so we can't make sure that each shard gets enough memory. This would probably be a bit complicated because we would have to make sure we're correctly tracking the memory used concurrently.

But now I see that we do actually know the number of shards in `smp::configure` before we `calculate_memory()`, so we could just adjust the calculations there and change the meaning of `reserve_additional_memory` to something like `reserve_additional_memory_per_shard`, especially because we added this option just for this use case (https://github.com/scylladb/seastar/pull/1328). This would require adjusting `runtime_info_table` after all.

I can start working on either of these approaches, but if you want, you can take care of it as well

**Comment by tchaikov** [comment]
@wmitros i don't have a strong preference on either of the approach. but yeah, it seems that the 2nd one is better, as long as we always execute the same set of udf on all shards. and since one cannot predict the number of shards, but she could understand the size of memory a regular udf could use. so a per-shard setting makes more sense even for non-scylladb developers. i will ping you with a patch on the Seastar side ASAP.

**Comment by tchaikov** [comment]
@wmitros see https://github.com/scylladb/seastar/pull/1462

**Pull request closed without merging**
